### ISSUE TITLE
observers/loggers should take recommendations into account

### ISSUE THREAD

**Issue by nikohansen** [numbbo/numbbo#169]
By default, all observers/loggers should log the recommendation calls. They log the same information as for a function call for the recommended solution, without increasing the function evaluation counter.

**Comment by ttusar** [comment]
How should the logger treat the non-existing y values of the recommended solutions?

**Comment by nikohansen** [comment]
It says "they log the same information as for a function call". What is the question then? What are the possible alternatives to choose from?

**Comment by nikohansen** [comment]
The proposition was that, by default, another line is added to the output which records the values for the recommended solution (same format as for a function call), while using the same number of evaluations (i.e. time stamp) as for the last recorded entry (i.e. evaluation or recommended solution).

**Comment by ttusar** [comment]
Sorry for that question, everything is clear now.

We have just discussed recommendations with @brockho and came to the following idea that covers many possible uses for recommendations. For example, the recommendations could be used for recommending solutions for the noisy problems, or if the algorithm uses some kind of model of the function it is optimizing, but also for giving the user the possibility to log more than what the chosen logger does (the user might want to output the current population at each generation, for example, and right now it can't do any such thing).

So, the idea is to log recommendations into separate files (different from the ones used for the regular logging). In addition to the evaluation number, y and x values, we could also log an additional string through which the user can output some additional information for each evaluation (we would need to change the interface a bit). We wouldn't set a limit to the number of recommendations the user wishes to log (more than one recommendation can be done simultaneously) and we would leave it to the post processing to interpret them.

**Comment by nikohansen** [comment]
I don't think we should (ab)use recommendations for anything else than what they are named for: to separate solution for which the solver wants to know a function value from solutions which the solver recommends as best current approximation to the optimum. Abusing this interface is on the long run quite likely to be more confusing than helpful and bug-prone. If the user wants to log something out of the ordinary which the logger can actually log, then the logger should "provide" a generic interface for doing so, e.g. a simple log-call.

**Comment by ttusar** [comment]
Good point, it would probably be better to provide a separate interface...

**Comment by ttusar** [comment]
Still, for the MO case, it makes sense to log the recommendations into a separate file (in the MO case you can decide to log the nondominated solutions only at the end or even to not log them at all).

**Comment by ttusar** [comment]
As agreed yesterday, the MO case should output recommendations into the same files as the nondominated solutions (i.e. if the nondominated solutions are not output, the recommendations are not output). However, to make this work correctly, the logger would need to delay output for one evaluation (the logger needs to wait to see if a recommendation comes after an evaluation).

**Comment by ttusar** [comment]
@nikohansen Just double-checking: we said that recommendation functions should have the following form, right:
`void coco_recommend_solution(coco_problem_t *problem, const double *x);`

**Comment by nikohansen** [comment]
Right, same as `void coco_evaluate_function(coco_problem_t *problem, const double *x, double *y);` but without return value.

**Comment by ttusar** [comment]
The change in the API is done, now the loggers "just" need to implement the functionality.

**Comment by nikohansen** [comment]
And the post-processing: it needs to ignore all but the last of multiple-lines with the same evaluations "time stamp".

**Comment by nikohansen** [comment]
> the logger needs to wait to see if a recommendation comes after an evaluation

If this is too intricate to implement, it seems comparatively simple to make the clean-up on the post-processing side.

**Comment by nikohansen** [comment]
Adding another perspective on recommendations: maybe `coco_recommend` should the only function the default observer interferes with, rather than `coco_evaluate_function`. The latter calls the former, thereby setting the default recommendation associated to the current evaluation. This seems to be a significant conceptual simplification, as is signifies that only recommendations are (and should be) related to the performance assessment (and hence only recommendations useful to be logged).

**Comment by nikohansen** [comment]
It turns out that using recommendations *only* is rather not viable in the noiseless scenario, see [here](http://lists.lri.fr/pipermail/bbob-discuss/2012-March/000192.html) and #955.

**Comment by nikohansen** [comment]
Rewinding from zero: 1) recommendations are written as evaluations but without touching the evaluation counter. 2) the post-processing reads in the evaluation (each first entry with a unique counter) and the recommendation (each last entry with a unique counter). The quality indicator used on recommendations is the one described for the noisy scenario #168. In the noise-less scenario, the better indicator value between recommendation and evaluation counts. 

Rationale: using only recommendations in the noise-free scenario, that is, recommendations *replace* evaluations, doesn't reflect reality, where we *can* pick the best from all *evaluated* solutions as a promising candidate for the practitioner. Using the last recommendation only is not desirable, because it produces a potentially non-monotonic performance measure with increasing number of evaluations. Using all recommendations opens the door to possible exploitation of the assessment specifics ("getting two shots instead of one" for each function call). Hence the compromise.

**Comment by nikohansen** [comment]
A possibly helpful viewpoint on the interface: the recommendation can be viewed as an optional argument to the evaluation call.

**Comment by nikohansen** [comment]
Another suggestion (figure out with @brockho): recommendations _do_ increase the evaluation counter but make the same number of following evaluations free. This simplifies the recommendation of a set of solutions (useful in the multi-objective case). It seems to complicate the case when one of the "free" evaluations hits a new target, in particular in the noiseless scenario.

**Comment by nikohansen** [comment]
Version 3 "improved":
- when a recommendation is evaluated: increment the evaluation counter _and a/the free-counter_ 
- when evaluating a new solution: decrement the free-counter and do not increment the evaluation counter iff the free-counter is larger than zero and the solution does not hit a new target

That's it.

**Comment by nikohansen** [comment]
The above approach still makes sense to me.

From the implementation perspective there is a simple alternative: we write recommendations in a different file just as a normal line and without touching the evaluation counter at all. The logic how to combine recombinations with evaluation is then postponed to reading the data in, instead of when writing them out. That seems to make even more sense to me and doesn't lock us in into any given resolution of combination.

**Comment by ttusar** [comment]
Interesting idea... I like that it's easy to implement in C :-) However, we would need to introduce yet another `?dat` file. Is it safe to assume the file name is the same as the corresponding `dat` file, just the extension is different or would we have to write it somewhere in the `info` file? 

The recommendations would be in separate files and therefore easier to discern from the evaluations by just looking at the files. Is this okay?

**Comment by nikohansen** [comment]
> Is it safe to assume the file name is the same as the corresponding dat file, just the extension is different or would we have to write it somewhere in the info file?

Same filename with different extension looks good to me. I don't think we need to pollute the info file with that.

> The recommendations would be in separate files and therefore easier to discern from the evaluations by just looking at the files. Is this okay?

Exactly. However, now I see a (small) potential problem: what when we are writing several runs in the same file (which I think we may do)?

Related: this also suggests to count the first start already as "restart", such that we always write the first line in the restart file and can then reconstruct the rest always unambiguously.

**Comment by ttusar** [comment]
> Exactly. However, now I see a (small) potential problem: what when we are writing several runs in the same file (which I think we may do)?

You mean restarts during a single run (1) or actually several runs (2)? In case of (1), we usually do not care when logging evaluations - why should we care for recommendations? In case of (2), we put a header line in between the runs, which tells them apart. Unless I'm misunderstanding something again :-)

**Comment by nikohansen** [comment]
> In case of (2), we put a header line in between the runs, which tells them apart.

I see, that should do it as well. I don't know whether it was always like that or whether the postprocessing does consider headers besides the one in the first line. Using headers as separators is also slightly more difficult to process in any case, because then the file data cannot be read in with a one-liner simply as an array while ignoring header lines.

**Comment by ttusar** [comment]
We are already using header lines to separate data from different runs (usually on different instances, but that doesn't have to be the case) in the `dat` files. It seems the most natural to keep this for restarts and recommendations too...

**Comment by ttusar** [comment]
Any wishes wrt to the extension used for the files with recommendations? (`rdat` is taken for restarts)

**Comment by nikohansen** [comment]
`mdat` or whatever else you prefer ;-)

**Comment by ttusar** [comment]
Hm... turns out that for the bi-objective case, supporting recommendations is quite a challenge, because it requires to either (1) create a copy of the AVL tree that contains the existing solutions and then adding the recommended solution to this copy or (2) only "virtually" add the recommended solution to the AVL tree (take it into account just for the sake of computing the indicator value, but not actually add it to the tree). Neither is appealing...

**Comment by nikohansen** [comment]
For the time being just write them to a file and ignore them otherwise? Does that make it impossible to integrate them afterwards?

**Comment by ttusar** [comment]
Thinking about this a bit more... Does it even make sense for the recommendation in the bi-objective case to log the indicator value (rather than the values of the objectives)? This has some meaning only in the context of other solutions residing in the archive at the same time.

**Comment by ttusar** [comment]
> For the time being just write them to a file and ignore them otherwise? Does that make it impossible to integrate them afterwards?

What do you mean by write them? Write their objective values?

**Comment by nikohansen** [comment]
I meant a regular output line, at least evaluation count and objective values.

**Comment by nikohansen** [comment]
Thinking about it, recommendations seem to make much less sense in the MO case, because all and arbitrary new points are non-dominated with high probability in the end (when the objectives locally strongly disagree).

**Comment by ttusar** [comment]
An update on the state of the code (on my fork, not yet on numbbo/coco):

The bbob (biobj) logger implements the `logger_bbob_recommend` (`logger_biobj_recommend`) function, which evaluates the solution and outputs information according to observer options to the file with recommendations. The evaluation result is not returned and the evaluation counter is not increased. 

So the functionality is there, but it is not using the `coco_recommend_solution` function of the `coco_problem`, which it should (would make the code cleaner). Also, since we are discussing that this does not make much sense in the MO case, we might want to remove that one.
